targets/mbedos5/tools/generate_pins.py

```diff
@@ -0,0 +1,224 @@
+    """
+    Walk through the directory tree, starting at root_dir, and enumerate all
+    valid include directories.
+    """
+    for root, dirs, files in os.walk(root_dir, topdown=True):
+        # modify dirs in place
+        dirs[:] = filter(lambda x: x in directory_labels 
+                                or (    not x.startswith('TARGET_')
+                                    and not x.startswith('TOOLCHAIN_')), dirs)
+        yield root
+
+
+class TypeDeclVisitor(c_ast.NodeVisitor):
+    def __init__(self, filter_names=[]):
+        self.names = filter_names
+
+    def visit(self, node):
+        value = None
+
+        if node.__class__.__name__ == "TypeDecl":
+            value = self.visit_TypeDecl(node)
+
+        if value is None:
+            for name, c in node.children():
+                value = value or self.visit(c)
+
+        return value
+
+    def visit_TypeDecl(self, node):
+        if node.declname in self.names:
+            c_gen = c_generator.CGenerator()
+            pins = {}
+
+            operators = DEFAULT_OPERATORS
+            operators[ast.BitOr] = lambda a, b: a | b
+            operators[ast.LShift] = lambda a, b: a << b
+            operators[ast.RShift] = lambda a, b: a << b
+            evaluator = SimpleEval(DEFAULT_OPERATORS )
+
+            for pin in node.type.values.enumerators:
+                expr = c_gen.visit(pin.value)
+
+                if "(int)" in expr:
+                    expr = expr.replace('(int)', '')
+
+                if expr in pins:
+                    pins[pin.name] = pins[expr]
+                else:
+                    pins[pin.name] = evaluator.eval(expr.strip())
+
+            return pins
+
+def enumerate_pins(c_source_file, include_dirs, definitions):
+    """
+    Enumerate pins specified in PinNames.h, by looking for a PinName enum
+    typedef somewhere in the file.
+    """
+    definitions += ['__attribute(x)__=', '__extension__(x)=', 'register=', '__IO=', 'uint32_t=unsigned int']
+
+    gcc_args = ['-E', '-fmerge-all-constants'] 
+    gcc_args += ['-I' + directory for directory in include_dirs] 
+
+    gcc_args += ['-D' + definition for definition in definitions]
+    ast = parse_file(c_source_file,
+                    use_cpp=True,
+                    cpp_path='arm-none-eabi-gcc',
+                    cpp_args=gcc_args,
+                    parser=GnuCParser())
+
+    # now, walk the AST
+    v = TypeDeclVisitor(['PinName'])
+    return v.visit(ast)
+
+
+if __name__ == "__main__":
+    if not os.path.exists('./mbed-os'):
+        print("Fatal: mbed-os directory does not exist.")
+        print("Try running 'make getlibs'")
+        sys.exit(1)
+
+    description = """
+    Generate pins.js for a specified mbed board, using target definitions from the
+    mbed OS source tree.
+    """
+
+    parser = argparse.ArgumentParser(description=description)
+
+    parser.add_argument('board', help='mbed board name')
+    parser.add_argument('-o',
```

Comment:
A friendlier name for the output files would be better imho. The `args.o` does not tell much by itself.